File: src/main/java/shopping/product/presentation/ProductController.java
package shopping.product.presentation;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import shopping.common.auth.AuthRole;
import shopping.common.auth.RoleRequired;
import shopping.common.dto.ApiResponse;
import shopping.product.application.dto.ProductCreateRequest;
import shopping.product.application.dto.ProductCreateResponse;
import shopping.product.application.dto.ProductGetResponse;
import shopping.product.application.dto.ProductUpdateRequest;
import shopping.product.application.command.*;

import java.util.List;
import shopping.product.application.dto.ProductUpdateResponse;
import shopping.product.application.query.ProductQueryService;

@RestController
@RequestMapping("/api/products")
@RequiredArgsConstructor
public class ProductController {

    private final ProductCreateUseCase productCreateUseCase;
    private final ProductQueryService productQueryService;
    private final ProductUpdateUseCase productUpdateUseCase;
    private final ProductDeleteUseCase productDeleteUseCase;

    @RoleRequired(AuthRole.ADMIN)
    @PostMapping
    public ResponseEntity<ApiResponse<ProductCreateResponse>> create(
        @Valid @RequestBody ProductCreateRequest request) {
        ProductCreateResponse response = productCreateUseCase.execute(request);
        return ResponseEntity.status(HttpStatus.CREATED)
            .body(ApiResponse.of(HttpStatus.CREATED, response));
    }

    @RoleRequired(AuthRole.ADMIN)
    @PatchMapping("/{productId}")
    public ResponseEntity<ApiResponse<ProductUpdateResponse>> update(
        @PathVariable Long productId,
        @Valid @RequestBody ProductUpdateRequest request) {
        ProductUpdateResponse response = productUpdateUseCase.execute(productId, request);
        return ResponseEntity.ok(ApiResponse.ok(response));
    }

    @RoleRequired(AuthRole.ADMIN)
    @DeleteMapping("/{productId}")
    public ResponseEntity<ApiResponse<Long>> delete(@PathVariable Long productId) {
        Long response = productDeleteUseCase.execute(productId);
        return ResponseEntity.ok(ApiResponse.ok(response));
    }

    @GetMapping("/{productId}")
    public ResponseEntity<ApiResponse<ProductGetResponse>> getById(
        @PathVariable Long productId) {
        ProductGetResponse response = productQueryService.getById(productId);
        return ResponseEntity.ok(ApiResponse.ok(response));
    }

    @GetMapping
    public ResponseEntity<ApiResponse<List<ProductGetResponse>>> getAll() {
        List<ProductGetResponse> response = productQueryService.getAll();
        return ResponseEntity.ok(ApiResponse.ok(response));
    }
}

File: src/main/kotlin/shopping/api/presentation/v1/ProductController.kt
package shopping.api.presentation.v1

import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.PutMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.ResponseStatus
import org.springframework.web.bind.annotation.RestController
import shopping.api.presentation.v1.dto.CreateProductRequest
import shopping.api.presentation.v1.dto.ProductResponse
import shopping.api.presentation.v1.dto.UpdateProductRequest
import shopping.api.presentation.v1.validator.ProductValidator
import shopping.application.ProductService
import shopping.support.response.ApiResponse

@RestController
@RequestMapping("/api/v1/products")
class ProductController(
    private val productService: ProductService,
    private val productValidator: ProductValidator,
) {
    @PostMapping
    @ResponseStatus(HttpStatus.CREATED)
    fun create(
        @RequestBody request: CreateProductRequest,
    ): ApiResponse<ProductResponse> {
        productValidator.validateCreate(request)
        val productId = productService.create(request.toCommand())
        val result = productService.getProduct(productId)
        return ApiResponse.success(ProductResponse.from(result))
    }

    @GetMapping("/{id}")
    fun getProduct(
        @PathVariable id: Long,
    ): ApiResponse<ProductResponse> {
        val result = productService.getProduct(id)
        return ApiResponse.success(ProductResponse.from(result))
    }

    @GetMapping
    fun getProducts(): ApiResponse<List<ProductResponse>> {
        val results = productService.getProducts()
        return ApiResponse.success(results.map { ProductResponse.from(it) })
    }

    @PutMapping("/{id}")
    fun update(
        @PathVariable id: Long,
        @RequestBody request: UpdateProductRequest,
    ): ApiResponse<Any> {
        productValidator.validateUpdate(request)
        productService.update(id, request.toCommand())
        return ApiResponse.success()
    }

    @DeleteMapping("/{id}")
    fun delete(
        @PathVariable id: Long,
    ): ApiResponse<Any> {
        productService.delete(id)
        return ApiResponse.success()
    }
}
